### DISCUSSION TITLE
CancelledError in uasyncio/stream.py in _serve when calling server.wait_closed()

### DISCUSSION THREAD

**Discussion by Beormund** [micropython/micropython#11410] [Using MicroPython]
I migrated to MicroPython v1.20.0 on 2023-04-26; Raspberry Pi Pico W with RP2040.

I'm getting an unhandled CancelledError when calling server.close followed by await server.wait_closed on the server object returned from await asyncio.start_server. 

File "uasyncio/stream.py", line 1, in wait_closed
File "uasyncio/core.py", line 1, in run_until_complete
File "uasyncio/stream.py", line 1, in _serve

I've set an exception handler on the loop which doesn't catch this CancelledError. Once this error is thrown all sockets created outside of start_server also become unresponsive, even if I wrap wait_closed in try/except.

**Comment by Josverl** [comment]
Did the same code work on an earlier build?
Can you share (a link to) the relevant section of the code, without that it's really hard to help

**Reply by Beormund** [reply]
I will put a small test demo together to demonstrate.

**Comment by peterhinch** [comment]
Looking at the `uasyncio` [code](https://github.com/micropython/micropython/blob/786013d467d8beace7c7a8a2cd1122c799d05dae/extmod/uasyncio/stream.py#LL129C1-L134C1) this is expected behaviour.
```py
    def close(self):
        self.task.cancel()

    async def wait_closed(self):
        await self.task
```
I don't think there's any need to issue `wait_closed()`. Task cancellation occurs the next time the task is scheduled, so in the following sequence shoud suffice:
```py
async def foo():
    server.close()
    await asyncio.sleep(0)
    # It's closed now
```

**Comment by Beormund** [comment]
Not sure why you get a CancelledError being thrown in stream.py _serve when the CancelledError is supposed to be caught:

```py
import uasyncio as asyncio
  
async def handle(r,w):
    pass
    
async def main():
    loop = asyncio.new_event_loop()
    server = await asyncio.start_server(handle, '0.0.0.0', 80)
    server.close()
    await server.wait_closed()
    loop.run_forever()
    
asyncio.run(main())

'''
Traceback (most recent call last):
  File "<stdin>", line 13, in <module>
  File "uasyncio/core.py", line 1, in run
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 10, in main
  File "uasyncio/stream.py", line 1, in wait_closed
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/stream.py", line 1, in _serve
CancelledError: '''
```

https://github.com/micropython/micropython/blob/786013d467d8beace7c7a8a2cd1122c799d05dae/extmod/uasyncio/stream.py#L135-L151

**Reply by peterhinch** [reply]
This code looks wrong to me. The commented lines should be deleted.
```py
import uasyncio as asyncio
  
async def handle(r,w):
    pass
    
async def main():
    # loop = asyncio.new_event_loop()  # Delete
    server = await asyncio.start_server(handle, '0.0.0.0', 80)
    server.close()
    await server.wait_closed()
    # loop.run_forever()  # Delete
    
asyncio.run(main())
```
The event loop stuff largely dates from old versions of Python's `asyncio` and I'm not sure of the implications of using them here
